src/interop/js-interop/usage.md

```diff
@@ -0,0 +1,454 @@
+
+When interacting with a JS value, you need to provide a Dart type for it. You
+can do this by either using or declaring an interop type. Interop types are
+either a ["JS type"] provided by Dart or an [extension type] wrapping an interop
+type. Interop types allow you to provide an interface for a JS value and lets
+you declare interop APIs for its members. They are also used in the signature
+of other interop APIs.
+
+```dart
+extension type Window(JSObject _) implements JSObject {}
+```
+
+`Window` here is an interop type for an arbitrary `JSObject`. There is no
+runtime guarantee that `Window` is actually a JS [`Window`]. See
+[the extension types documentation] for more info on the runtime semantics of
+extension types. There also is no conflict with any other interop interface
+that is defined for the same value. If you do want to check that `Window` is
+actually a JS `Window`, you can [check the type of the JS value through
+interop].
+
+You can also declare your own interop type for the JS types Dart provides by
+wrapping them or an interop type they implement with no conflict:
+
+```dart
+extension type Array._(JSArray<JSAny?> _) implements JSArray<JSAny?> {
+  external Array();
+}
+```
+
+In most cases, you will likely declare an interop type using `JSObject` as the
+[representation type] because you're likely interacting with JS objects which
+don't have an interop type provided by Dart.
+
+Interop types should also generally [implement] their representation type so
+that they can be used where the representation type is expected, like in
+[`package:web`].
+
+## Interop members
+
+[`external`] interop members provide an idiomatic syntax for JS members. They
+allow you to write a Dart type signature for any arguments and return value.
+The types that can be used in the signature of these members have
+[restrictions]. The JS API they correspond to is determined by a combination of
+where they're declared, their name, what kind of Dart member they are, and any
+[renames](#js).
+
+### Top-level interop members
+
+Given the following JS members:
+
+```JS
+globalThis.name = 'global';
+globalThis.isNameEmpty = function() {
+  return globalThis.name.length == 0;
+}
+```
+
+You can write interop members for them like so:
+
+```dart
+@JS()
+external String get name;
+
+@JS()
+external set name(String value);
+
+@JS()
+external bool isNameEmpty();
+```
+
+Here, there exists a property `name` and a function `isNameEmpty` that are
+exposed in the global scope. To access them, you use top-level interop members.
+To get and set `name`, you declare and use an interop getter and setter with the
+same name. To use `isNameEmpty`, you declare and call an interop function with
+the same name. You can declare top-level interop getters, setters, methods, and
+fields, which is the same as a getter and setter pair.
+
+Top-level interop members must be declared with a [`@JS()`](#js) annotation to
+distinguish them from other `external` top-level members.
```

Comment:
nit: consider saying ", for example, from `dart:ffi`" -- I'm not sure if users may be aware what other external members there may be, so not sure if it's worth giving an example.

Reply:
Done.